File: backend/app/services/cache_service.py
import hashlib
import json
import logging
from typing import Optional

logger = logging.getLogger(__name__)


class CacheService:

    def __init__(self):
        self._cache: dict = {}
        self._redis = None
        self._init_redis()

    def _init_redis(self):
        try:
            import redis
            from app.config import get_settings
            settings = get_settings()
            self._redis = redis.from_url(settings.redis_url, decode_responses=True)
            self._redis.ping()
            logger.info("Redis cache connected")
        except Exception as e:
            logger.warning(f"Redis unavailable, using in-memory cache: {e}")
            self._redis = None

    def _make_key(self, prefix: str, data: str) -> str:
        hash_val = hashlib.md5(data.encode()).hexdigest()
        return f"{prefix}:{hash_val}"

    async def get_query_cache(self, query: str, document_ids: list[str]) -> Optional[dict]:
        key = self._make_key("query", f"{query}:{sorted(document_ids)}")

        if self._redis:
            cached = self._redis.get(key)
            if cached:
                logger.info(f"Cache HIT for query: {query[:50]}...")
                return json.loads(cached)
        else:
            if key in self._cache:
                return self._cache[key]

        return None

    async def set_query_cache(self, query: str, document_ids: list[str], result: dict, ttl: int = 3600):
        key = self._make_key("query", f"{query}:{sorted(document_ids)}")
        value = json.dumps(result, default=str)

        if self._redis:
            self._redis.setex(key, ttl, value)
        else:
            self._cache[key] = result

    async def invalidate_document_cache(self, document_id: str):
        if self._redis:
            try:
                keys = self._redis.keys("query:*")
                if keys:
                    self._redis.delete(*keys)
            except Exception as e:
                logger.warning(f"Cache invalidation failed: {e}")
        else:
            query_keys = [k for k in self._cache if k.startswith("query:")]
            for key in query_keys:
                del self._cache[key]


cache_service = CacheService()

File: backend/src/services/cacheService.js
'use strict';
const crypto = require('crypto');
const config = require('../config');

class CacheService {
  constructor() {
    this._mem = new Map();
    this._redis = null;
    this._initRedis();
  }

  _initRedis() {
    try {
      const Redis = require('ioredis');
      this._redis = new Redis(config.redisUrl, {
        lazyConnect: true,
        enableOfflineQueue: false,
        // Disable automatic retries — we fall back to memory instead
        maxRetriesPerRequest: 0,
        retryStrategy: () => null,
        reconnectOnError: () => false,
      });
      this._redis.on('connect', () => console.log('[Cache] Redis connected'));
      this._redis.on('error', () => {}); // suppress ioredis default error log
      this._redis.connect().catch((err) => {
        console.warn(`[Cache] Redis unavailable (${err.message}) — using in-memory cache`);
        this._redis = null;
      });
    } catch {
      console.warn('[Cache] ioredis not available — using in-memory cache');
    }
  }

  _key(query, docIds) {
    const raw = `${query}:${JSON.stringify([...docIds].sort())}`;
    return 'query:' + crypto.createHash('md5').update(raw).digest('hex');
  }

  async get(query, docIds) {
    const key = this._key(query, docIds);
    if (this._redis) {
      try {
        const val = await this._redis.get(key);
        if (val) { console.log(`[Cache] HIT: ${query.slice(0, 50)}`); return JSON.parse(val); }
      } catch { /* fall through */ }
    }
    return this._mem.get(key) || null;
  }

  async set(query, docIds, result, ttl = 3600) {
    const key = this._key(query, docIds);
    const val = JSON.stringify(result);
    if (this._redis) {
      try { await this._redis.setex(key, ttl, val); return; } catch { /* fall through */ }
    }
    this._mem.set(key, result);
    // simple TTL for in-memory
    setTimeout(() => this._mem.delete(key), ttl * 1000);
  }

  async invalidateDocument() {
    if (this._redis) {
      try {
        const keys = await this._redis.keys('query:*');
        if (keys.length) await this._redis.del(...keys);
        return;
      } catch { /* fall through */ }
    }
    for (const k of this._mem.keys()) {
      if (k.startsWith('query:')) this._mem.delete(k);
    }
  }
}

module.exports = new CacheService();